### ISSUE TITLE
Formatting the data for 'tsfresh'

### ISSUE THREAD

**Issue by Nuri8** [blue-yonder/tsfresh#213]
First off, thank you for this amazing library, which showed me another way of observing data.

I read the tutorials, but I think I clearly don't get it right somewhere...
Below is the DataFrame 'df' I was trying to feed into 'tsfresh': multiple tickers with n 'features' so that the machine can learn the 'label' for the particular ticker on the particular date.

![a](https://user-images.githubusercontent.com/29172610/27144540-3ee23354-516d-11e7-97ee-6d678a611468.png)

```
X = extract_features(df[:, :-1], column_id='ticker', column_sort='date')
y = df['label']
```

But then, [ValueError: Index of X must be a subset of y's index] occurred in that the number of unique 'column_id' equals to 3 whereas the number of 'label' equals to 15. I know this is what the tutorial explains; each robot will be predicted 'only once' with all the corresponding time-series data of features.

My intention was to predict each 'label' per 'time step' for each 'ticker' like the figure above. Could you please help me out with this?

**Comment by MaxBenChrist** [comment]
Hi @Nuri8,

the features_1 ... features_n are denoting the value of your time series? From those you want to extract the features?

If yes, you have to omit the first two columns. Also, right now, your data is not in the right format. Please see http://tsfresh.readthedocs.io/en/latest/text/data_formats.html

**Comment by Nuri8** [comment]
![axxx](https://user-images.githubusercontent.com/29172610/27166402-45d7bd74-51d5-11e7-924d-032e729528cb.png)

I believe this might work in order to feed into 'tsfresh', please correct me if I am wrong.

1. Allocate an identical 'ID' to a particular 'date', so that there will be only one 'label' for the same 'ID' (and consequently, there will be only one 'label' for the same 'date').
2. Since I have several time-series values ranging from feature_1 to feature_n (sensors in the tutorial) for each 'ID', stack them within the same 'ID' by assigning each time-series value to a 'column_kind'. This will help me further in case I want to roll the time-series values for each 'column_kind'.
3. Do 'extract_features' like the following:
```
X = extract_features(df[:, :-1], 
                       column_id='id', 
                       column_sort='date', 
                       column_kind='column_kind', 
                       column_value='TS values')
y = df['label']
```
I think this is the intended usage of 'tsfresh' for 'one' robot. But what if I have 'multiple' robots that I want to predict? The modified table above was only for 'AAPL'. If I have other tickers to extract features, do I have to perform the whole 'extract_features' process individually for each ticker?

**Comment by MaxBenChrist** [comment]
There is one slight error: 

```
X = extract_features(df[:, :-1], 
                       column_id='id', 
                       column_sort='date', 
                       column_value='TS values')
y = df['label']
```
You only have one type of the time series, the TS values.

**Comment by nils-braun** [comment]
To answer your second question: you could for example introduce another index, where you concatonate the id and the ticker name and than extract the features synchronously for all tickers and all IDs.

**Comment by MaxBenChrist** [comment]
I think this question again shows that we need a forecasting example, preferable with stock data.

**Comment by fereshtemohebbi** [comment]
@nils-braun  Hi Could you explain more about your comment I dont understand exactly what should we have to do in this situation ?

**Comment by nils-braun** [comment]
Because TSFresh in the moment only allows one single column as the ID (and will probably also continue to do so) - so having two columns which both are needed to define the information as the ID are too much for our API.

**Comment by MaxBenChrist** [comment]
We should not support multiple id columns. concatenation of existing id columns into one is trivial with pandas. I will close this

**Comment by wsjzha** [comment]
Hello, can you explain further about how to format the data? Take Nuri8's post as an example,

![image](https://user-images.githubusercontent.com/5770525/27924880-cf51af7a-6250-11e7-8991-44bdc004b8bd.png)

Say the data is not related to stock, but it's collected from two sensors monitoring two events. On 2008-01-03,  event A happened and I observed n samples (corresponds to n features in the table above) from sensor A and B. Here, I collected n/2 samples from each sensor. On 2008-01-04, event B happened and I observed n samples from both sensor. On 2008-01-07, event A happened again and n samples are observed.

I know time is only used for sorting, but I am confused here,
1. considering the new example, should I use the exact format as shown in Nuri8's post? If I should, then what's the difference between n samples from sensor A and B together and n samples from sensor A only?
2. should we seperate the samples from different sensors and then extract features from the samples of the same sensor within the same day? I believe using the corrent data format, we will mix all the samples from both sensors and extract the feature from there.

Hopefully my description is clear. Thanks for any help!

**Comment by Ethiral** [comment]
I am in the same situation.@wsjzha Did you figure out any effective solution to your third point.
Thanks

**Comment by sskulki** [comment]
I am still facing issues with extracting the features using tsfresh. The solution didn't work @Nuri8  and @MaxBenChrist. Also, I have another kind of problem one of the column contains list of integers, as shown below:
date                col1             col3
2-2-2010    [1,2,3,...]     [4,1,4,1...]
2-2-2011    [1,6,7,3,...]     [8,2,4,1...]
........
Ay idea how to extract the features from this kind of dataframe using tsfresh?
-S

**Comment by kjchoi10** [comment]
Anyone solve this?

**Comment by nils-braun** [comment]
Hm, I think I did not understand this clearly. Maybe there is some misunderstanding on what tsfresh is doing, so let my try to clarify that:
* tsfresh is a feature extraction library for time series. This means, for every time series (= a list of single numbers, sorted by some kind of "time" (does not need to be real time)) we extract a bunch of features, which are also just numbers.
  Now I try to map this to your example, but I somehow fail to do this. What is your time series? If you say, one sensor sees N samples, is this your time series? Or are you interested in the series over several days?
  Please note that so far this has nothing to do with classification, but just with extracting features so far.
* Each time series has only a single value over time - tsfresh has no multi-column features so we just treat them separately. 
* In the nomenclature of tsfresh, each specific time series has a specific id.
* If you want to do forecasting, its basically the same. You just create multiple time series out of a single timeseries (by cutting/cropping repeatedly) - each of them will have a different ID.
* Once you have defined what your time series is, you can extract features for each of those series and then do the classification. So you need one label for each time series, which will produce one set of features, which you can then use for training.

Hope this helps. Maybe you can try to map your data into the format of different time series for which you want to have features, giving each time series a specific ID, sort column and column name (which we call "kind").

**Comment by Marble90** [comment]
Hi everyone,

I have a problem with reordering my data frame to a form that tsfresh can understand it. There are 2 different links that make me confused:
first the explanation in :
Features dataframe with tsfresh is totally different from the thing described in :
blue-yonder/tsfresh#213
In the first link, one feature from timestep 0 to N is considered as a separate id. BUT in the second link, feature 1 to N in a single timestep is considered as a separate id. Which one is correct?

My case is classifying a time series data frame in which I have only 2 labels (1 and 0) as my target. If I follow the first link, after feature_extraction(), I won't get the equal number of ids (rows) and labels(target) at the end and I won't be able to use the feature_selection() afterward. However, I am not sure about the description in the second link.

Would you plz help me?

**Comment by nils-braun** [comment]
Hi @Marble90! 
What might confuse you, is that tsfresh understands multiple input formats. They are all described in https://tsfresh.readthedocs.io/en/latest/text/data_formats.html

Now concerning your use case: did I understand correctly that you have a label for every time step? That sound like a windowing/rolling use case. Have a look into https://tsfresh.readthedocs.io/en/latest/text/forecasting.html. 
For more help, could you maybe share some pieces of your data? Does not need to be the real data - but to understand better how the schema looks like.

**Comment by Marble90** [comment]
Thank you SO much for your reply @nils-braun ,

For every timeStep (rows), I have a label. I have 2 hours of rig data extracted from a simulator and each column represents a parameter (feature) I want to do the feature engineering on them.
Here is a summarized part of my dataframe:
<img width="550" alt="dataframe" src="https://user-images.githubusercontent.com/54379250/105721487-ecb9e800-5f24-11eb-804a-970e2702e0d5.png">

Except 'TimeStep' and 'label' I have around 20 different parameters, but here I only show 6 of them for simplicity which are: PressureBottomHole, Hookload, and so on.
According to the link [https://stackoverflow.com/questions/62071901/how-to-use-tsfresh-python-package-to-extract-features-from-time-series-data](), Firstly, I dropped 2 columns 'TimeStep' and 'label', and follow the code
`time = df_ts['TimeStep']
y = df_ts['label']
df_ts.drop( 'TimeStep','lable', axis = 1, inplace = True )
df_ts_list = df_ts.values.tolist()

df_ts_id = pd.DataFrame()
for i, ts in enumerate(df_ts_list):
    data = [[x, i] for x in ts]
    df_ts_id = df_ts_id.append(data, ignore_index=True)
df_ts_id.columns = ['value', 'id']
`
The outcome would be:
<img width="143" alt="reOrdered" src="https://user-images.githubusercontent.com/54379250/105721787-3aceeb80-5f25-11eb-8a4f-3785cee6b722.png">

Then if I run the code below:
`with warnings.catch_warnings():
    warnings.simplefilter( "ignore" )
    df_ts_features = extract_features(df_ts_id, column_id='id')`
I would get a dataframe of 10 rows × 779 columns in which the columns are NEW features I guess. Then I can append the 'label' column again and search for most "relevant" features based on `extract_relevant_features(df, label, column_id='id')`

BUT according to this link [https://stackoverflow.com/questions/44679191/features-dataframe-with-tsfresh]()
for this dataframe (3 features are provided this one):
<img width="301" alt="dataframe2" src="https://user-images.githubusercontent.com/54379250/105723855-8c787580-5f27-11eb-9887-59d2c35e376f.png">
if I run the following code:
`y = df_ts['label']
df_ts.drop( 'label', axis = 1, inplace = True )

df_ts.index.rename([ 'time', 'id'], inplace = True ) 
df_ts = df_ts.reset_index()`

The result would be something different from the previous approach:
<img width="228" alt="reOrdered2" src="https://user-images.githubusercontent.com/54379250/105724229-eda04900-5f27-11eb-8a0d-637f9cf132b0.png">

Exactly like the issue @Nuri8 discussed, after feature extraction, I have a dataframe of 3 rows × 779 columns and I CANNOT append my 10-rows 'label' and choose the most "relevant" ones.
`with warnings.catch_warnings():
    warnings.simplefilter( "ignore" )
    f = extract_features( df_ts, column_id = "id",column_sort = 'time' )`

So, this really confused me and I am still struggling with this since last December. My final purpose is to predict when 'label' is turned to "1" and which features (parameters) affect the most on this event.  

Sorry for the long explanation. I just want to clarify.
Thanks

**Comment by nils-braun** [comment]
Thank you @Marble90 for your detailed explanation. Now I understand your problem much better and can hopefully help you.
I will make an assumption here (you can tell me if it is correct): for predicting the label of each row, you do not only need the data of this specific row but also the data from the past (because if this is not the case and the data for this specific row is sufficient, you do not have a time series problem but actually a very "normal" ML classification problem).

Now, tsfresh is a time series feature extraction library. So you need to give tsfresh your time series. I would encourage you to read through https://tsfresh.readthedocs.io/en/latest/text/data_formats.html to get the connection. Bet let me try to map this to your data.
There are 3 columns which are of special interest to tsfresh.
* the time (or sort). That one is easy, it is the "TimeStep".
* the value(s). That one is also easy, because it is just all your features. You do not need to give this to tsfresh in this case (as it is just everything).
* the id. This column separates your time series from each other and is the complicated one.

Now the question is: what are your time series? Let's come back to my assumption: What I would propose to use to e.g. predict the label at t = 599 is to use all data (with all 20 features) of the last 10 timesteps (how many to use depends on your use case, you can change this). That is, as I have mentioned, a rolling use case. Rolling means that you produce out of a single time series from 1 to 10, separate smaller time series of shorter length, e.g. 1-5, 2-6, 3-7, ... You can then use the features extracted from the time series 1-5 to predict the label at 5, the one from 2-6 for 6 etc.
Have a look into https://tsfresh.readthedocs.io/en/latest/text/forecasting.html for more information.
Mapped to your data:

```python
import pandas as pd
import tsfresh as ts
from tsfresh import extract_features
from tsfresh.utilities.dataframe_functions import roll_time_series

# a smaller shorter version of your data
df = pd.DataFrame({
                    "TimeStep": [1, 2, 3, 4, 5, 6, 7, 8],
                    "PressureBottomHole": [4, 5, 6, 7, 8, 9, 10, 11],
                    "HookLoad": [4, 5, 6, 7, 8, 9, 10, 11],
                  })

# important, add a dummy column
df["id"] = 1

# roll
df_rolled = roll_time_series(df, column_id="id", column_sort="TimeStep", 
                             max_timeshift=10)

# extract
df_features = extract_features(df_rolled, column_id="id", column_sort="TimeStep")
```

If you now compare: `df_features` has 8 rows, and so has `df`. One row consists of all features calculated using the data 10 time steps in the past up until now. You can now map this directly with your labels again and continue your ML pipeline.

Please, definitely have a look into our documentation to understand more about the different techniques. The code you have shared is all valid - but for very different use cases (where the data is not structured like your data). Before you use tsfresh, you need to understand what actually your time series is - otherwise you will have a hard time mapping those concepts.

I hope that helped!

**Comment by Marble90** [comment]
@nils-braun  Thanks a million!
Although I had read the [ https://tsfresh.readthedocs.io/en/latest/text/data_formats.html](url), I was totally unaware of the magic of [https://tsfresh.readthedocs.io/en/latest/text/forecasting.html](url). It really did the trick. Thanks.

YOU ARE RIGHT. I am working with time-series forecasting and ML modeling and for a prediction, I need data from the past as well. 

***From many features to a few that are useful***
Now the df_features dataframe can be concatenated with "label" column.
According to this [https://tsfresh.readthedocs.io/en/latest/api/tsfresh.feature_selection.html](url), I tried to pick out only few columns (out of 5000 different columns) from df_features to prepare my dataset for train/test splitting. 

with this code:
`df_selected = select_features(df_features, y, fdr_level =0.05)`
I got an error:
<img width="691" alt="seelct_features" src="https://user-images.githubusercontent.com/54379250/106024826-748c2780-60c8-11eb-9bb7-6bd670b3aab3.png">

I played with fdr_level with extremely low values (and with extremely high, just to double check) but I have the same index error. 

I searched for another solution to get rid of irrelevant features and I found:
`y = df['label']`
`df_relevant = extract_relevant_features(df_rolled, y, column_id="id", column_sort="TimeStep")`

But of course df_rolled is not equal with y in the number of rows. So I changed it to:
`df_relevant = extract_relevant_features(df_features, y)`
and delete the column_id="id", column_sort="TimeStep" as they don't exist in df_features. No success !

I was thinking maybe there is another approach toward time-series "feature_selection" in the tsfresh documentation.
After generating df_features with thousands of features, how do we select the best and most relevant features that have statistical relationship with the "label" variable?

**Comment by nils-braun** [comment]
I think you are not very far away from the solution. Have a look into the rolled dataframe and especially into the ids of it. They have a very specific format. This format also needs to be the index (pandas index) of your y. If this is the case, you can use `select_features` as you have described. 

You have three possibilities: either you edit the index of y, or you replace the id column of the rolled dataframe (so that it fits y) or you treat y "as another column" before rolling (so that it also gets rolled just as all other normal columns) and split it out before feature extraction (with a group by on id and the selecting always the last entry). 

If you have problems with that, I can also provide some code examples. Please also have a look into our notebook examples in the repo, especially on the stock market prediction notebooks (similar, although not completely the same problem)

**Comment by Marble90** [comment]
IF I am wrong please correct me. 
since now, I have a dataframe named "df_rolled" which has an "id" column with a specific format. 
Also, I have another dataframe named "df_features" which is generated by:
`df_features = extract_features(df_rolled, column_id="id", column_sort="TimeStep")`

***NOW***
Out of 3 possibilities you named above, if I want to use `df_selected = select_features(df_features, y, fdr_level =0.05)` for select the most relevant features and y is my final target for classification : `y = df['label']` , 
---> is this line enough/correct to do so? 

`y.index = df_features.index`

Because when I applied this, I got only 7 columns of features (out of 5000 features) and I guess this is the prepared database I should work on it for the ML modeling.

**Comment by nils-braun** [comment]
>  is this line enough/correct to do so?
Looks reasonable, yes

>  I got only 7 columns of features
That is not a lot, but could be ok (depending on how easy it is to actually fo the prediction). Just a comment: if you are using ML afterwards which is robust against "too many features" (such as a BDT), you might want to just pass on all the features.

> Can I assign different "id" (for example 2,3, etc) to my other simulations:
Yes, that is correct!

>  In the robot example, how do we interpret the "kind" property?
A timeseries (with one id) is all the data, which belongs to the same "entity" (robot in this case). Different kinds are the different variables (e.g. different sensors) that belong to this entity (in the robot use case, it is the different F_x, F_y, F_z etc). As you have data in the so-called "wide-format", you should not specify kind at all (all your columns except the id and the time are of type "kind"). 

> Why did you suggest using rolling(windowing) for my use case?
The idea behind time series regression or classification is, that you have for each timeseries (could be one dimension or multi dimensional, e.g. one kind or multiple kinds) you have a single label you want to train on. In your example above, that would mean that for all your 100 rows with id = 2 (and all the columns, which are the kinds) you only have a single label that you want to predict. And for other 100 rows (with a different id), you have another single label.
Now you explained to me in some post above that this is not the case. For your 100 rows you have 100 labels. This means we first need to re-create this relation timeseries -> single label (not timeseries -> multiple labels). And this is exactly what rolling does. 

You might get confused because of the different "time series" objects that we are talking about. A time series in general is a series of numbers (with a time column) that could be one-dimensional or multi-dimensional (multi in your case, as you have many kind columns: "Pressure", SurfaceTorque" etc.). Each of those separate, distinct time series have a distinct id (could be e.g. measurements on a different day, measurements of different objects etc.). 
As I said, ML on time series always works in the way: you have many different time series (with different id) and for each timeseries (each id) you have a single label. As this is not the case for you, we "artificially" need to generate new time series so that the number of time series and the number of labels is the same. Let me give you an example: we start with two time series (with id = 0 and id = 1), both having 100 entries in two columns (Pressure and Torque). Instead of having two labels you now have 200 labels. So how do we get 200 time series out of our 2? We use rolling: for each label, we extract the data of this time series until this label. That will generate 100 time series out of a single time series. To keep our mantra "separate time series, separate id", we now give each of those new time series a new id (in tsfresh, we call it "(1, 20)" if they come from the previous id = 1 and include the data until row 20). 

I would invite you to try out the rolling and the feature extracting with some small test data (e.g. [1, 2, 3]) and see what happens.

**Comment by Marble90** [comment]
Your explanation is awesome and It becomes all clear. 

My data is generated by a simulator and the final goal is to predict when the label is changed to 1 from zero (disaster happening in the rig). Each simulation starts in TimeStep=0 and continues for some amount of time (not necessarily equal). Since controlling this disaster (suppressing the kick) is not that easy by the simulator, I could only create one or at most 2 kicks and stop the simulation and had to start a new one, again. However, the primary parameters in equal in all the simulations.

If I execute the code `extract_features()` and `select_features()` merely on my first simulation, the result will be a dataframe consists of 7 columns. If I concatenate 2 simulations and execute the code, the result will be a different 11-column dataframe with different column names. 

Maybe you wonder why don't I deal with datasets separately and then merge them for ML? Because for a precise ML modeling, I need to have enough observations to dedicate 70% of observations to the **training** and the rest to the **testing**.
As I mentioned above, with a simulator I couldn't generate many disasters (10 or even more) to be able to do the forecasting on one single time-series dataset. Instead, I have to manage many time-series datasets each of which has one or 2 disasters in different points in time: 
<img width="925" alt="MLmodeling" src="https://user-images.githubusercontent.com/54379250/107636094-70394000-6c6c-11eb-9abd-14dc6ffa83e8.png">
<img width="889" alt="MLmodeling2" src="https://user-images.githubusercontent.com/54379250/107636096-70d1d680-6c6c-11eb-849f-5d224ef88c68.png">

To have enough observations (kicks), the only thing is to merge these datasets. But
1. should I concatenate them all, and then execute the `extract_features()` and `select_features()`? 
Or `extract_features()` and `select_features()` should be done independently on them?
2. if I merge all the time-series data into one time-series dataset, this data will not be sequential in time. Then how would the training/testing splitting work? How can I handle my ML modeling in respect of time?

**Comment by nils-braun** [comment]
To 1: use as much data as possible for the selection process (for the extraction it does not matter as the extraction happens independently on each block anyways). So first concatenate, then select.
To 2: I am not sure if I understand that question. If the order is a problem, you could just sort them. I assume you are thinking about a look-ahead-bias? The data after rolling is still sorted in time (both in ach of the rolling windows as well as the rolling windows among each other). It does not matter of the data for one simulation comes before the data of another simulation as there is no dependency between them. You could for example train your model on the union of the first 70% of each simulation and then test it on the remaining 30%.

**Comment by rodinael** [comment]
Hello everyone, I have problem while using tsfresh. I have 2- ID(1001 &1002) with time series dataset for eye measurements. I extract features from right_pupil so I get only one extracted value for each ID. So now I have two problems:

 First my target is numerical variable so I have same length like the original series before extracting the features so how can I calculate_relevance_table because extracted_features and y dose not have the same length now.                  

second problem I have two variables I can use it as indexes which is  ID and trials so each subject(ID) have 140 trial how to extract the features for subject per trial.

<img width="817" alt="my dataseet" src="https://user-images.githubusercontent.com/49314807/163487519-a68d1d16-5288-48ef-9cce-7d22960d2446.png">

My codes:

    
from tsfresh import extract_features
from tsfresh.feature_selection.relevance import calculate_relevance_table
from tsfresh.utilities.dataframe_functions import impute


X=extracted_features = extract_features(
    x_test,
    column_id="subject",
    column_sort="time",
    column_value="right_pupil",
    # we impute = remove all NaN features automatically
                     impute_function=impute)
relevance_table = calculate_relevance_table(extracted_features, y)
relevance_table = relevance_table[relevance_table.relevant]
relevance_table.sort_values("p_value", inplace=True)
print(relevance_table)
